Improvements to CPUMAP usage based on production experience

netoptimizer opened the pull request:
This PR is two fairly simple commits that solve issues in production.

Details on the issues solved are described in details in the commit message.

netoptimizer committed:
xdp_iphash_to_cpu: Pin and share cpumap between all interfaces

Experience from production shows that it is sub-optimal to run several
CPUMAP kthread's on the same CPU as they will be competing for sched
runtime and cause more context-switches.

A single CPUMAP *can* be shared between several concurrent XDP
programs, because softirq makes sure that two RX-NAPI instances
doesn't run concurrently on the *same* CPU (and NIC drivers calls
xdp_do_flush on their NAPI exit, which flush map state).

Thus, this patch pin and and share cpumap between all interfaces.

The 'cpus_available' is not pinned, but kept tied to the XDP program
loaded on the interface. This allows uses of xdp_iphash_to_cpu to
control what CPUs are available per interface basis.

Signed-off-by: Jesper Dangaard Brouer <brouer@redhat.com>

netoptimizer committed:
cpumap qsize: allow changing and set new larger default

Experience from production shows queue size of 192 is too small, as
this caused packet drops during cpumap-enqueue on RX-CPU.  This can be
diagnosed with xdp_monitor sample program.

This bpftrace program was used to diagnose the problem in more detail:

 bpftrace -e '
  tracepoint:xdp:xdp_cpumap_kthread { @deq_bulk = lhist(args->processed,0,10,1); @drop_net = lhist(args->drops,0,10,1) }
  tracepoint:xdp:xdp_cpumap_enqueue { @enq_bulk = lhist(args->processed,0,10,1); @enq_drops = lhist(args->drops,0,10,1); }'

Watch out for the @enq_drops counter. The @drop_net counter can happen
when netstack gets invalid packets, so don't despair it can be
natural, and that counter will likely disappear in newer kernels as it
was a source of confusion (look at netstat info for reason of the
netstack @drop_net counters).

The production system was configured with CPU power-saving C6 state.
Learn more in this blogpost[1].

And wakeup latency in usec for the states are:

 # grep -H . /sys/devices/system/cpu/cpu0/cpuidle/*/latency
 /sys/devices/system/cpu/cpu0/cpuidle/state0/latency:0
 /sys/devices/system/cpu/cpu0/cpuidle/state1/latency:2
 /sys/devices/system/cpu/cpu0/cpuidle/state2/latency:10
 /sys/devices/system/cpu/cpu0/cpuidle/state3/latency:133

Deepest state take 133 usec to wakeup from (133/10^6). The link speed
is 25Gbit/s ((25*10^9/8) in bytes/sec). How many bytes can arrive with
in 133 usec at this speed: (25*10^9/8)*(133/10^6) = 415625 bytes. With
MTU size packets this is 275 packets, and with minimum Ethernet (incl
intergap overhead) 84 bytes it is 4948 packets. Clearly default queue
size is too small.

Setting default cpumap queue to 2048 as worst-case (small packet) at
10Gbit/s is 1979 packets with 133 usec wakeup time, +64 packet before
kthread wakeup call (due to xdp_do_flush) worst-case 2043 packets.

Thus, if a packet burst on RX-CPU will enqueue packets to a remote
cpumap CPU that is in deep-sleep state it can overrun the cpumap queue.

The production system was also configured to avoid deep-sleep via:
 tuned-adm profile network-latency

[1] https://jeremyeder.com/2013/08/30/oh-did-you-expect-the-cpu/

Signed-off-by: Jesper Dangaard Brouer <brouer@redhat.com>

tohojo commented on src/xdp_iphash_to_cpu_user.c:
```diff
@@ -532,6 +543,9 @@ int main(int argc, char **argv)
 			add_all_cpus = true;
 			break;
+		case 's':
+			qsize = strtoul(optarg, NULL, 0);
```
Check the return value, maybe? :)

netoptimizer replied:
Well, I would have to check `errno` for invalid or out-of-range input values.
I was thinking, the kernel will reject this if it is out-of-range.
Looking at manual, I can see that it can return 0 on `EINVAL`, which will be bad actually.

netoptimizer replied:
Notice on kernel side qsize==0 is same as deleting cpumap entry, which is why is it extra bad

netoptimizer reviewed (commented):
To myself @netoptimizer please add check against qsize zero.

netoptimizer committed:
xdp_iphash_to_cpu_user: Validate qsize input

On invalid input to strtoul() the value zero can be returned.
This qsize value is dangerous as the kernel will delete the
cpumap entry when giving qsize zero.

Fixes: 582edcb7026f ("cpumap qsize: allow changing and set new larger default")
Signed-off-by: Jesper Dangaard Brouer <brouer@redhat.com>

netoptimizer committed:
xdp_iphash_to_cpu_user: Use errno if bpf_map_update_elem fails

For example if using --qsize larger than 16384 we see:

 Create CPU entry failed err(75):Value too large for defined data type

Signed-off-by: Jesper Dangaard Brouer <brouer@redhat.com>

tohojo reviewed (approved):
LGTM
